### PR TITLE
Fix #12332: Avoid closure creation

### LINKED ISSUE #12332: Misleading example about "Postconditions" on page about context functions
https://dotty.epfl.ch/docs/reference/contextual/context-functions.html#example-postconditions says:

> Since WrappedResult is an opaque type alias, its values need not be boxed, and since ensuring is added as an extension method, its argument does not need boxing either. Hence, the implementation of ensuring is as about as efficient as the best possible code one could write by hand:
> ```scala
> val s =
>   val result = List(1, 2, 3).sum
>   assert(result == 6)
>   result
> ```

"Is about as efficient" doesn't seem to be true since it still allocates and executes a closure. The only extra feature it adds is offering the `ensuring(result == ...)` syntax. You can get exactly the same cost using `def ensuring(condition: T => Boolean): T = { assert(condition(x)); x }` and calling it `.ensuring(_ == 6)`.

If you want less cost you should use `inline` (which works both for the regular syntax and the `result == ` one if you remove the opaque type).

### PR DESCRIPTION
**Pull request by liufengyun**
Fix #12332: Avoid closure creation

### CODE REVIEW

File: docs/docs/reference/contextual/context-functions.md [hunk 1]
```diff
@@ -117,12 +117,12 @@ As a larger example, here is a way to define constructs for checking arbitrary p
 
 ```scala
 object PostConditions:
-   opaque type WrappedResult[T] = T
+   type WrappedResult[T] = T
```
Comment (smarter) [inline docs/docs/reference/contextual/context-functions.md:120]: If it's just a type alias then WrappedResult[T] might as well be removed in favor of T, but I believe the point of using an opaque type was to make sure the `result` method below always returned the parameter coming from `ensuring` and not some random other implicit (e.g. in `1.ensure { given Int = 2; result == 1 }`, you want `result` to still be `1`.

**Review by smarter — CHANGES_REQUESTED** [review]

File: docs/docs/reference/contextual/context-functions.md (hunk 1 above)
Comment (liufengyun) [reply docs/docs/reference/contextual/context-functions.md:120]: Yes, good point. If we have to pay a cost for the additional safety, then we can remove `zero-overhead abstraction` in the text.

Comment (liufengyun) [reply docs/docs/reference/contextual/context-functions.md:120]: BTW, the safety is not bullet-proof -- it's possible to break it by confusing two `WrappedResult[Int]` via a mutable `var`. we have to rely on end-users to write reasonable code here.

Comment (odersky) [reply docs/docs/reference/contextual/context-functions.md:120]: The idea is that `WrappedResult` is implementation internal. Use code should not mention it. In that case it is bullet-proof, I think.

**Comment by odersky** [comment]
We definitely need to keep the type opaque. That's explained in the text. So we should weaken the performance claims.

File: docs/docs/reference/contextual/context-functions.md (hunk 1 above)
Comment (liufengyun) [reply docs/docs/reference/contextual/context-functions.md]: If we assume users cannot write the type explicitly in programmers, it's harder to break it. But it's still possible to do so, given that we have type inference and `export` mechanism.

**Comment by liufengyun** [comment]
> We definitely need to keep the type opaque. That's explained in the text. So we should weaken the performance claims.

Now we updated the document instead.

**Review by OlivierBlanvillain — APPROVED** [review]

File: docs/docs/reference/contextual/context-functions.md [hunk 2]
```diff
@@ -145,6 +145,9 @@ val s =
   assert(result == 6)
   result
 ```
+
+The only additional cost if creation of the function at runtime. However, we
```
Comment (liufengyun) [inline docs/docs/reference/contextual/context-functions.md:149]:
```suggestion
The only additional cost is the creation of the function at runtime. However, we
```

**Comment by liufengyun** [comment]
This is outdated.

**Pull request closed without merging**